Breaking: Spawn webserver in separate process

molant opened the pull request:
<!--

Read our pull request guide:
https://webhint.io/docs/contributor-guide/getting-started/pull-requests/

For the following items put an "x" between the square brackets
(i.e. [x]) if you completed the associated item.

-->

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

This currently spawns a new server per test. I was thinking about doing something smart and have just one per test file but ava spawns every test in parallel so I'm not sure that would change anything.

Also this PR is missing:

* [x] Update the testing documentation on how the test server works
* [x] Update the `create-hint` package

But it will be good if you can start taking a look into it. 

Fix #1694

molant commented:
BTW, I passed all the tests on Windows for the first time in a very long time 😊

molant commented on package.json:
```diff
@@ -54,6 +54,7 @@
     "pre-release": "npm run clean:packages && npm run build:scripts && npm run release -- --prerelease",
     "release": "npm run clean:packages && npm run build:scripts && node dist/scripts/release.js",
     "test": "npm run clean && node scripts/test.js",
+    "test-hints": "loop \"yarn test-only\" --cwd packages --exclude configuration-development,configuration-progressive-web-apps,configuration-web-recommended, connector-chrome, connector-edge, connector-jsdom, connector-local, create-hint, create-hintrc, create-parser, extension-browser, extension-vscode, formatter-codeframe, formatter-excel, formatter-html, formatter-json, formatter-stylish, fo ...
```
I'll remove this.

molant commented on packages/utils-tests-helpers/src/hint-runner.ts:
```diff
@@ -124,31 +124,26 @@ const validateResults = (t: ExecutionContext<HintRunnerContext>, results: Proble
-    test.afterEach.always(async (t) => {
-        await t.context.server.stop();
-    });
+    // /**
```
Will delete this.

molant commented:
@sarvaje I've done your feedback

molant commented:
* [x] Update the `create-hint` package ➡ Doesn't need any manual update. We mark this as a breaking change so the release script updates everything.

@alrra, did you change the script so major version bumps of `devDependencies` are patches/minors?

sarvaje commented:
@molant why are you ignoring `connector-chrome` in so many tests?

sarvaje commented on packages/utils-create-server/README.md:
```diff
@@ -9,3 +9,50 @@ To install the package, you need to run:
+* Better isolation
+* Improve the debugging experience
+
+With the current architecture lnowing what is actually returned by the
```
```suggestion
With the current architecture knowing what is actually returned by the
```

molant commented:
> why are you ignoring `connector-chrome` in so many tests?

Windows build keep failing on Windows and I'm testing things to make it work. I was expecting them to be more reliable by spawning less chrome instances :(

sarvaje commented:
> Windows build keep failing on Windows and I'm testing things to make it work. I was expecting them to be more reliable by spawning less chrome instances :(

But it is just for testing, or is it going to be permanent?

molant commented:
> But it is just for testing, or is it going to be permanent?

🤷‍♂️, I'm doing tests for now

alrra commented:
>     * [x]  Update the `create-hint` package ➡ Doesn't need any manual update. We mark this as a breaking change so the release script updates everything.
> 
> 
> @alrra, did you change the script so major version bumps of `devDependencies` are patches/minors?

Working on it, will be done later today, or Monday.

For  `create-hint` there shouldn't be any problems, because nothing depends on it.

molant commented:
I think this is ready for merge. I've done a few changes since the last review so it will be cool if you could review it again:
- Fix a race condition with CDP that was causing the Windows failures (that's what I believe, Windows builds are now passing on Pipelines 🎉)
- Refactor `requestResponse` in jsdom to match the one in chrome (forgot about it last time, both files should look about the same now)
- group some `await`s into `await Promise.all()`

Also this PR should be rebased instead of squashed. I can do it later today if I get enough 👍

antross commented on packages/utils-connector-tools/tests/requester.ts:
```diff
@@ -206,9 +204,9 @@ const loopServerMultiSteps = {
+    const { requester } = t.context;
 
-    server.configure(hopsServerConfig);
+    const server = await Server.create({ configuration: hopsServerConfig });
```
Does this need a corresponding `await server.stop()`? I'm only marking this one, but there are multiple occurances in the file.

molant replied:
yes, that's why it was failing on Linux and mac at the moment. I've pushed the changes

molant commented on packages/utils-create-server/src/server.ts:
```diff
@@ -0,0 +1,505 @@
+w4C7/uFcTns/5rYjv+vp+EGv5cupxJjXpd1O/buT/Tt/Ur5maySbu4vDYVgmPuXU
+7avEZJTeYgih
+-----END CERTIFICATE-----`
+    private static readonly _key: string =
```
Perf, that way we avoid reading from disk and it's not like we are going to change them.

sarvaje replied:
Ok!

molant commented:
> Looks good to me with the updates assuming tests pass.

Tests pass! mac is failing because of #1772 which happens randomly but that's unrelated to this PR.
